rllib/agents/impala/impala.py

```diff
@@ -230,6 +236,23 @@ def validate_config(config):
             "`num_aggregation_workers` should be significantly smaller than"
             "`num_workers`! Try setting it to 0.5*`num_workers` or less.")
 
+    # If two separate optimizers/loss terms used for tf, must also set
+    # `_tf_policy_handles_more_than_one_loss` to True.
+    if config["_separate_vf_optimizer"] is True:
+        # Only supported to tf so far.
+        # TODO(sven): Need to change APPO|IMPALATorchPolicies (and the models
+        #  to return separate sets of weights in order to create the different
+        #  torch optimizers).
+        if config["framework"] not in ["tf", "tf2", "tfe"]:
+            raise ValueError(
+                "`_separate_vf_optimizer` only supported to tf so far!")
+        if config["_tf_policy_handles_more_than_one_loss"] is False:
```

Comment:
"is there ever a case where we have 1 optimizer but multiple loss function?"
No, but this is an existing API that we would change if we didn't carefully introduced this new flag. The user has to explicitly set this flag to True to agree to the API change. Otherwise, we would break code for existing users of our TFPolicy API.
I think this is a typical path we should go whenever we change an API:
1) Flag explicitly set to allow for the new API to be used.
2) Set this flag to True by default, catch possible errors for those users that are still on the old API, and warn if flag is left False.
3) Error if flag is left to False.
4) Remove flag and all checks for it in the code: Behave always as if flag was set to True.

Reply:
No, this has nothing to do with the model builder API. It's a plain change in the TFPolicy API. Note also that the TorchPolicy API is already behaving like this (flag is True): You can return more than one loss term from the loss function, but then also need the same number of optimizers returned from your optimizer function.

So we are just unifying TFPolicy and TorchPolicy behavior (but we have to do this carefully as to not break existing code!).